packages/ceramic-common/src/doctype.ts

```diff
@@ -28,142 +28,144 @@ export enum AnchorStatus {
+  [index: string]: any; // allow arbitrary properties
 }
 
 /**
  * Document params
  */
 export interface DocParams {
-    metadata?: DocMetadata;
+  metadata?: DocMetadata;
 
-    [index: string]: any; // allow arbitrary properties
+  [index: string]: any; // allow arbitrary properties
 }
 
 /**
  * Document information about the next iteration
  */
 export interface DocNext {
-    content?: any;
-    owners?: Array<string>;
-    metadata?: DocMetadata;
+  content?: any;
+  owners?: Array<string>;
+  metadata?: DocMetadata;
 }
 
 /**
  * Document state
  */
 export interface DocState {
-    doctype: string;
-    content: any;
-    next?: DocNext;
-    metadata: DocMetadata;
-    signature: SignatureStatus;
-    anchorStatus: AnchorStatus;
-    anchorScheduledFor?: number; // only present when anchor status is pending
-    anchorProof?: AnchorProof; // the anchor proof of the latest anchor, only present when anchor status is anchored
-    log: Array<CID>;
+  doctype: string;
+  content: any;
+  next?: DocNext;
+  metadata: DocMetadata;
+  signature: SignatureStatus;
+  anchorStatus: AnchorStatus;
+  anchorScheduledFor?: number; // only present when anchor status is pending
+  anchorProof?: AnchorProof; // the anchor proof of the latest anchor, only present when anchor status is anchored
+  log: Array<CID>;
 }
 
 /**
  * Doctype init options
  */
 export interface DocOpts {
-    applyOnly?: boolean;
-    skipWait?: boolean;
+  applyOnly?: boolean;
+  skipWait?: boolean;
 }
 
 /**
  * Describes common doctype attributes
  */
 export abstract class Doctype extends EventEmitter {
-    constructor(private _state: DocState, private _context: Context) {
-        super()
-    }
-
-    get id(): string {
-        return DoctypeUtils.createDocIdFromGenesis(this._state.log[0])
-    }
-
-    get doctype(): string {
-        return this._state.doctype
-    }
-
-    get content(): any {
-        return cloneDeep(this._state.content)
-    }
-
-    get metadata(): DocMetadata {
-        return cloneDeep(this._state.metadata)
-    }
-
-    get owners(): Array<string> {
-        return cloneDeep(this._state.metadata.owners)
-    }
-
-    get head(): CID {
-        return this._state.log[this._state.log.length - 1]
-    }
-
-    get state(): DocState {
-        return cloneDeep(this._state)
-    }
-
-    set state(state: DocState) {
-        this._state = state
-    }
-
-    set context(context: Context) {
-        this._context = context
-    }
-
-    get context(): Context {
-        return this._context
-    }
-
-    /**
-     * Makes a change on an existing document
-     * @param params - Change parameters
-     * @param opts - Initialization options
-     */
-    abstract change(params: DocParams, opts?: DocOpts): Promise<void>
+  constructor(private _state: DocState, private _context: Context) {
+    super()
+  }
+
+  get id(): string {
+    return DoctypeUtils.createDocIdFromGenesis(this._state.log[0])
```

Comment:
I usually try to avoid performing computations on getters because it's not obvious what is happening from the consuming side.

For example the consumer could have code using `doctype.id` multiple times and each time it makes this computation, while if the method was called `getId()` the consumer could instead do `const id = doctype.getId()` and avoid calling it multiple times.

An alternative could also be something like this:

```ts
get id(): string {
  if (this._id == null) {
    this._id = DoctypeUtils.createDocIdFromGenesis(this._state.log[0])
  }
  return this._id
}
```
In cases a getter is likely to be used multiple times and the computed value is immutable, it could be cached on the instance on the first access.

Reply:
Note that this getter was not changed by this PR.

Reply:
Yes no need to change this now I think, I just commented on it as I noticed it.